# THE Impact Rankings 2025
- https://www.timeshighereducation.com/impactrankings
- part : #!/length/-1/sort_by/rank/sort_order/asc/cols/scores
- https://www.timeshighereducation.com/impactrankings#!/length/-1/sort_by/rank/sort_order/asc

In [40]:
## URLS
#urls = ["https://www.timeshighereducation.com/world-university-rankings/2023/regional-ranking/", "https://www.timeshighereducation.com/world-university-rankings/2024/regional-ranking/", "https://www.timeshighereducation.com/world-university-rankings/2025/regional-ranking"]
#part1 = '#!/length/-1/sort_by/rank/sort_order/asc/cols/scores/'
#urls[0] + part1
url1 ='https://www.timeshighereducation.com/impactrankings'
url ='https://www.timeshighereducation.com/impactrankings#!/length/-1/sort_by/rank/sort_order/asc'
url

'https://www.timeshighereducation.com/impactrankings#!/length/-1/sort_by/rank/sort_order/asc'

In [ ]:
import requests
import urllib
from bs4 import BeautifulSoup
response = requests.get('http://google.com')
assert response.status_code < 400

In [ ]:
#function check if url exists
import httpx
from bs4 import BeautifulSoup

def get_title_httpx(url):
    try:
        with httpx.Client(timeout=5) as client:
            response = client.get(url)
            if response.status_code == 200:
                soup = BeautifulSoup(response.text, 'html.parser')
                return soup.title.string.strip() if soup.title else "No title found"
            else:
                return f"Status code: {response.status_code}"
    except httpx.RequestError as e:
        return f"Error: {e}"

In [ ]:
print(get_title_httpx("https://www.python.org"))
print(get_title_httpx(url))

## install libraries if required

In [8]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager
from selenium.webdriver.common.by import By
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException
import time
import pandas as pd
import numpy as np
import re

In [9]:
pd.set_option('max_colwidth',None)

In [10]:
# Setup Chrome options
options = Options()
options.add_argument("--start-maximized")
options.add_argument("--disable-notifications")

In [11]:
# Download location
prefs = { "download.prompt_for_download": False, "download.default_directory": "/Users/du/dup/auData/download"}
options.add_experimental_option("prefs", prefs)

In [12]:
cp2 = '/opt/homebrew/bin/chromedriver'  # Replace with the actual path to your chromedriver

In [ ]:
#start driver
cService = webdriver.ChromeService(executable_path=cp2)
driver = webdriver.Chrome(service = cService)
driver.get(url)
wait = WebDriverWait(driver, 3)
print(driver.title)

In [ ]:
# Maximize window and wait for page to load
driver.maximize_window()
time.sleep(5)  
driver.find_element(By.ID,'CybotCookiebotDialogBodyButtonDecline').click()

In [ ]:
year = re.search(r'\d{4}', driver.title).group()
print(driver.title, 'Year - ' , year)

In [ ]:
#check if table element exists
table = driver.find_element(By.ID,'datatable-1')
table

In [ ]:
table.text[0:1000]

In [ ]:
# Locate the table header row
header_elements = driver.find_elements(By.CSS_SELECTOR, '#datatable-1 thead th')
# Extract and print header text
headers = [header.text.strip() for header in header_elements]
print(headers)

In [ ]:
# Another way to download Headings
thead = driver.find_element(By.CSS_SELECTOR, '#datatable-1 thead')
thead
th_elements = thead.find_elements(By.TAG_NAME, "th")
th_elements
header_list = [th.text.strip() for th in th_elements]
header_list

## HTML STRUCTURE
<tbody>
    <tr>
        <td class="rank sorting_1 sorting_2">1</td>
        <td class="name namesearch">
            <a class="ranking-institution-title" href="/world-university-rankings/western-sydney-university" data-position="title" data-mz="">Western Sydney University</a>
            <div class="location"><span><a href="/location/a4zw0000000GnyFAAS">Australia</a></span></div> </td>
        <td class="best-scores">
            <div class="sdg-score-multi">
                <div class="sdg-score-multi_item">
                    <div class="sdg-score-multi_number">15</div>
                    <div class="sdg-score-multi_value-wrapper">
                        <div class="sdg-score-multi_value"> <p>95.1</p> </div> </div></div>
                 <div class="sdg-score-multi_item">
                    <div class="sdg-score-multi_number">12</div>
                    <div class="sdg-score-multi_value-wrapper">
                        <div class="sdg-score-multi_value"><p>92.6</p> </div> </div></div>       
                 <div class="sdg-score-multi_item">
                    <div class="sdg-score-multi_number">11 </div>
                    <div class="sdg-score-multi_value-wrapper">
                        <div class="sdg-score-multi_value"><p>92.2</p> </div> </div></div>         
                <div class="sdg-score-multi_item">
                    <div class="sdg-score-multi_number">17 </div>
                    <div class="sdg-score-multi_value-wrapper">
                        <div class="sdg-score-multi_value"><p>98.1</p> </div> </div></div>
            </div>
        </td>
        <td class=" scores overall-score">98.7</td>
    </tr>
</tbody>

### Selectors
- row //*[@id="datatable-1"]/tbody
- rank th[aria-label='Rank: activate to sort column ascending']

In [132]:
# Step 1: Find all rows inside tbody
rows = driver.find_elements(By.CSS_SELECTOR, "table#datatable-1 tbody tr")
rows[0:5] #if 5 elements are shown

[<selenium.webdriver.remote.webelement.WebElement (session="3d8a5caac536fcfcfb6ba5aba3a2cb45", element="f.45772AD9ADD03D76F52CC2E0B279BE3D.d.249C47E654DEE146905C34F3D376FFCA.e.1774")>,
 <selenium.webdriver.remote.webelement.WebElement (session="3d8a5caac536fcfcfb6ba5aba3a2cb45", element="f.45772AD9ADD03D76F52CC2E0B279BE3D.d.249C47E654DEE146905C34F3D376FFCA.e.1652")>,
 <selenium.webdriver.remote.webelement.WebElement (session="3d8a5caac536fcfcfb6ba5aba3a2cb45", element="f.45772AD9ADD03D76F52CC2E0B279BE3D.d.249C47E654DEE146905C34F3D376FFCA.e.1700")>,
 <selenium.webdriver.remote.webelement.WebElement (session="3d8a5caac536fcfcfb6ba5aba3a2cb45", element="f.45772AD9ADD03D76F52CC2E0B279BE3D.d.249C47E654DEE146905C34F3D376FFCA.e.1775")>,
 <selenium.webdriver.remote.webelement.WebElement (session="3d8a5caac536fcfcfb6ba5aba3a2cb45", element="f.45772AD9ADD03D76F52CC2E0B279BE3D.d.249C47E654DEE146905C34F3D376FFCA.e.1776")>]

In [155]:
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
len(rows2)

2232

In [156]:
sdg_container = rows2[1].find_element(By.CSS_SELECTOR, "td.best-scores .sdg-score-multi")
sdg_container
sdg_items = sdg_container.find_elements(By.CLASS_NAME, "sdg-score-multi__item")
sdg_items  #4 items
rank = rows2[1700].find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
name = rows2[1700].find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
country = rows2[1700].find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
print(rank, name, country)

1501+ ICFAI University, Dehradun India


In [157]:
sdg_scores = {}
sdg_container = rows2[1].find_element(By.CSS_SELECTOR, "td.best-scores .sdg-score-multi")
sdg_container
sdg_items = sdg_container.find_elements(By.CLASS_NAME, "sdg-score-multi__item")
sdg_items  #4 items
for item in sdg_items:
        number = item.find_element(By.CLASS_NAME, "sdg-score-multi__number").text.strip()
        value = item.find_element(By.CLASS_NAME, "sdg-score-multi__value p").text.strip()
        sdg_scores[number] = value
        print(number, value)

11 95.3
15 95.9
12 93.5
17 94.5


In [183]:
all_data = []
i=0
#print(rows[0:1])
for row in rows2:
    i = i + 1
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    # Find SDG container *within this row*
    sdg_container = row.find_element(By.CSS_SELECTOR, "td.best-scores .sdg-score-multi")
    # Now find each SDG item within that container
    sdg_items = sdg_container.find_elements(By.CLASS_NAME, "sdg-score-multi__item")
    sdg_scores = {}
    for item in sdg_items:
        number = item.find_element(By.CLASS_NAME, "sdg-score-multi__number").text.strip()
        value = item.find_element(By.CLASS_NAME, "sdg-score-multi__value p").text.strip()
        sdg_scores[number] = value
        #print(number, value)
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    #print(i, rank, name)
    #print('Rank is ', rank, ', Name is', name, ', Country is ', country, ', Overall Score - ', overall, ' Other Scores ', sdg_scores) #, name, country)   
    all_data.append({   "rank": rank, "university": name, "country": country,  "sdg_scores": sdg_scores, "overall_score": overall })
    #all_data.append({   "rank": rank, "university": name, "country": country,  "overall_score": overall })
print("No of rows extracted ", i)

No of rows extracted  2232


In [184]:
type(all_data)

list

In [185]:
len(all_data)

2232

In [187]:
sdata= all_data[0:2]
sdata

[{'rank': '1',
  'university': 'Western Sydney University',
  'country': 'Australia',
  'sdg_scores': {'15': '95.1', '12': '92.6', '11': '92.2', '17': '98.1'},
  'overall_score': '98.7'},
 {'rank': '2',
  'university': 'University of Manchester',
  'country': 'United Kingdom',
  'sdg_scores': {'11': '95.3', '15': '95.9', '12': '93.5', '17': '94.5'},
  'overall_score': '98.4'}]

In [188]:
# Print or convert to DataFrame
for record in all_data[0:2]:    print(record)

{'rank': '1', 'university': 'Western Sydney University', 'country': 'Australia', 'sdg_scores': {'15': '95.1', '12': '92.6', '11': '92.2', '17': '98.1'}, 'overall_score': '98.7'}
{'rank': '2', 'university': 'University of Manchester', 'country': 'United Kingdom', 'sdg_scores': {'11': '95.3', '15': '95.9', '12': '93.5', '17': '94.5'}, 'overall_score': '98.4'}


In [189]:
sdg_scores_list = [entry['sdg_scores'] for entry in sdata if 'sdg_scores' in entry]
for i, scores in enumerate(sdg_scores_list, 1):
    print(f"Entry {i} SDG Scores: {scores}")

Entry 1 SDG Scores: {'15': '95.1', '12': '92.6', '11': '92.2', '17': '98.1'}
Entry 2 SDG Scores: {'11': '95.3', '15': '95.9', '12': '93.5', '17': '94.5'}


In [190]:
flat_data = []
for entry in all_data:
    base = {k: v for k, v in entry.items() if k != 'sdg_scores'}
    if 'sdg_scores' in entry:
        for i, (sdg_no, sdg_val) in enumerate(entry['sdg_scores'].items(), start=1):
            base[f'SDG{i}_no'] = sdg_no
            base[f'SDG{i}_value'] = sdg_val
    flat_data.append(base)

df = pd.DataFrame(flat_data)
print(df.shape)

(2232, 12)


In [232]:
df1 = df.copy()

In [233]:
df1.head()

rank                           university         country overall_score  \
0    1            Western Sydney University       Australia          98.7   
1    2             University of Manchester  United Kingdom          98.4   
2    3  Kyungpook National University (KNU)     South Korea          97.8   
3   =4                  Griffith University       Australia          97.7   
4   =4               University of Tasmania       Australia          97.7   

  SDG1_no SDG1_value SDG2_no SDG2_value SDG3_no SDG3_value SDG4_no SDG4_value  
0      15       95.1      12       92.6      11       92.2      17       98.1  
1      11       95.3      15       95.9      12       93.5      17       94.5  
2       8       92.0       9       99.5      15       91.6      17       96.8  
3      15       95.6       6       91.8      14       94.3      17       96.4  
4      15       96.1      13       93.2      14       90.1      17       96.2

In [234]:
# Step 1: Add serial number
df1.insert(0, 'ser', range(1, len(df) + 1))
df1.head()

ser rank                           university         country  \
0    1    1            Western Sydney University       Australia   
1    2    2             University of Manchester  United Kingdom   
2    3    3  Kyungpook National University (KNU)     South Korea   
3    4   =4                  Griffith University       Australia   
4    5   =4               University of Tasmania       Australia   

  overall_score SDG1_no SDG1_value SDG2_no SDG2_value SDG3_no SDG3_value  \
0          98.7      15       95.1      12       92.6      11       92.2   
1          98.4      11       95.3      15       95.9      12       93.5   
2          97.8       8       92.0       9       99.5      15       91.6   
3          97.7      15       95.6       6       91.8      14       94.3   
4          97.7      15       96.1      13       93.2      14       90.1   

  SDG4_no SDG4_value  
0      17       98.1  
1      17       94.5  
2      17       96.8  
3      17       96.4  
4      17       96.2

In [235]:
df1.head()

ser rank                           university         country  \
0    1    1            Western Sydney University       Australia   
1    2    2             University of Manchester  United Kingdom   
2    3    3  Kyungpook National University (KNU)     South Korea   
3    4   =4                  Griffith University       Australia   
4    5   =4               University of Tasmania       Australia   

  overall_score SDG1_no SDG1_value SDG2_no SDG2_value SDG3_no SDG3_value  \
0          98.7      15       95.1      12       92.6      11       92.2   
1          98.4      11       95.3      15       95.9      12       93.5   
2          97.8       8       92.0       9       99.5      15       91.6   
3          97.7      15       95.6       6       91.8      14       94.3   
4          97.7      15       96.1      13       93.2      14       90.1   

  SDG4_no SDG4_value  
0      17       98.1  
1      17       94.5  
2      17       96.8  
3      17       96.4  
4      17       96.2

In [236]:
# Step 2: Clean 'rank'
def clean_rank(val):
    if pd.isna(val): return np.nan
    val = re.sub(r'[^\d\-–]', '', str(val))  # remove non-digit/range chars
    if '-' in val or '–' in val:
        parts = re.split(r'[-–]', val)
        try:
            return np.mean([int(p.strip()) for p in parts if p.strip().isdigit()])
        except:
            return np.nan
    return int(val)

df1['rank2'] = df1['rank'].apply(clean_rank)
df1['rank2'] = df1['rank2'].fillna(-1).astype(int)
# Step 3: Clean 'overall_score'
def clean_score(val):
    if pd.isna(val): return np.nan
    val = str(val)
    if '-' in val or '–' in val:
        parts = re.split(r'[-–]', val)
        try:
            return np.mean([float(p.strip()) for p in parts])
        except:
            return np.nan
    try:
        return float(val)
    except:
        return np.nan

df1['overall_score2'] = df1['overall_score'].apply(clean_score)

In [237]:
df1.head()

ser rank                           university         country  \
0    1    1            Western Sydney University       Australia   
1    2    2             University of Manchester  United Kingdom   
2    3    3  Kyungpook National University (KNU)     South Korea   
3    4   =4                  Griffith University       Australia   
4    5   =4               University of Tasmania       Australia   

  overall_score SDG1_no SDG1_value SDG2_no SDG2_value SDG3_no SDG3_value  \
0          98.7      15       95.1      12       92.6      11       92.2   
1          98.4      11       95.3      15       95.9      12       93.5   
2          97.8       8       92.0       9       99.5      15       91.6   
3          97.7      15       95.6       6       91.8      14       94.3   
4          97.7      15       96.1      13       93.2      14       90.1   

  SDG4_no SDG4_value  rank2  overall_score2  
0      17       98.1      1            98.7  
1      17       94.5      2            98.4  
2      17       96.8      3            97.8  
3      17       96.4      4            97.7  
4      17       96.2      4            97.7

In [193]:
#driver.refresh()
#driver.quit()

In [238]:
# Step 1: Identify the columns to melt
id_vars = ['ser','rank','rank2', 'university', 'country', 'overall_score','overall_score2']
value_vars = [col for col in df1.columns if col.startswith('SDG')]
# Step 2: Melt the dataframe to long format
df_melted = df1.melt(id_vars=id_vars, value_vars=value_vars, var_name='sdg_type', value_name='sdg_info')
df_melted

ser   rank  rank2                             university  \
0         1      1      1              Western Sydney University   
1         2      2      2               University of Manchester   
2         3      3      3    Kyungpook National University (KNU)   
3         4     =4      4                    Griffith University   
4         5     =4      4                 University of Tasmania   
...     ...    ...    ...                                    ...   
17851  2228  1501+   1501                       Zarqa University   
17852  2229  1501+   1501     Zaytoonah International University   
17853  2230  1501+   1501  Zewail City of Science and Technology   
17854  2231  1501+   1501  Zhytomyr Ivan Franko State University   
17855  2232  1501+   1501      Ziane Achour University of Djelfa   

              country overall_score  overall_score2    sdg_type   sdg_info  
0           Australia          98.7            98.7     SDG1_no         15  
1      United Kingdom          98.4            98.4     SDG1_no         11  
2         South Korea          97.8            97.8     SDG1_no          8  
3           Australia          97.7            97.7     SDG1_no         15  
4           Australia          97.7            97.7     SDG1_no         15  
...               ...           ...             ...         ...        ...  
17851          Jordan      1.7–49.7            25.7  SDG4_value  63.3–69.9  
17852           Syria      1.7–49.7            25.7  SDG4_value   2.1–44.7  
17853           Egypt      1.7–49.7            25.7  SDG4_value  44.8–58.0  
17854         Ukraine      1.7–49.7            25.7  SDG4_value   2.1–44.7  
17855         Algeria      1.7–49.7            25.7  SDG4_value   2.1–44.7  

[17856 rows x 9 columns]

In [239]:
# Step 3: Create group key for each SDG set
df_melted['sdg_group'] = df_melted['sdg_type'].str.extract(r'(\d+)')  # Extract the number from column name
df_melted['sdg_field'] = df_melted['sdg_type'].str.extract(r'(no|value)')  # no or value

# Step 4: Pivot back to one row per SDG with separate columns for no and value
df_final = df_melted.pivot_table(index=id_vars + ['sdg_group'], 
                                 columns='sdg_field', 
                                 values='sdg_info',
                                 aggfunc='first').reset_index()

# Step 5: Rename columns for clarity
df_final = df_final.rename(columns={'no': 'SDG_no', 'value': 'SDG_value'})

# Optional: Sort if needed
df_final = df_final.sort_values(['rank', 'SDG_no'])

# Preview the result
df_final.head()

sdg_field  ser       rank  rank2                 university    country  \
2            1          1      1  Western Sydney University  Australia   
1            1          1      1  Western Sydney University  Australia   
0            1          1      1  Western Sydney University  Australia   
3            1          1      1  Western Sydney University  Australia   
3949       988  1001–1500   1250         Al-Ayen University       Iraq   

sdg_field overall_score  overall_score2 sdg_group SDG_no  SDG_value  
2                  98.7            98.7         3     11       92.2  
1                  98.7            98.7         2     12       92.6  
0                  98.7            98.7         1     15       95.1  
3                  98.7            98.7         4     17       98.1  
3949          49.8–60.8            55.3         2      1  57.6–62.4

In [241]:
# Step 3: Clean 'overall_score'
def clean_score(val):
    if pd.isna(val): return np.nan
    val = str(val)
    if '-' in val or '–' in val:
        parts = re.split(r'[-–]', val)
        try:
            return np.mean([float(p.strip()) for p in parts])
        except:
            return np.nan
    try:
        return float(val)
    except:
        return np.nan

df_final['SDG_value2'] = df_final['SDG_value'].apply(clean_score)

In [124]:
df_final.head()

NameError: name 'df_final' is not defined

In [243]:
df_final.to_csv("/Users/du/dup/auData/ranking/the/theimpact25b.csv", index=False)

In [245]:
df_final[['ser','rank2','university', 'country','overall_score2', 'SDG_no','SDG_value2']].head(10)

sdg_field   ser  rank2                      university    country  \
2             1      1       Western Sydney University  Australia   
1             1      1       Western Sydney University  Australia   
0             1      1       Western Sydney University  Australia   
3             1      1       Western Sydney University  Australia   
3949        988   1250              Al-Ayen University       Iraq   
3954        989   1250             Al-Azhar University      Egypt   
3985        997   1250           Al-Nahrain University       Iraq   
3998       1000   1250            Al Turath University       Iraq   
4000       1001   1250  American University of Nigeria    Nigeria   
4030       1008   1250       Antonio Nariño University   Colombia   

sdg_field  overall_score2 SDG_no  SDG_value2  
2                    98.7     11       92.20  
1                    98.7     12       92.60  
0                    98.7     15       95.10  
3                    98.7     17       98.10  
3949                 55.3      1       60.00  
3954                 55.3      1       49.35  
3985                 55.3      1       55.20  
3998                 55.3      1       42.60  
4000                 55.3      1       60.00  
4030                 55.3      1       42.60

## SDG Wise

In [94]:
# SDG Goals Wise
sdglist = ['https://www.timeshighereducation.com/impactrankings','https://www.timeshighereducation.com/impactrankings/no-poverty', 'https://www.timeshighereducation.com/impactrankings/zero-hunger', 'https://www.timeshighereducation.com/impactrankings/good-health-and-well-being',
'https://www.timeshighereducation.com/impactrankings/quality-education', 'https://www.timeshighereducation.com/impactrankings/gender-equality', 'https://www.timeshighereducation.com/impactrankings/clean-water-and-sanitation', 
 'https://www.timeshighereducation.com/impactrankings/affordable-and-clean-energy', 'https://www.timeshighereducation.com/impactrankings/decent-work-and-economic-growth', 'https://www.timeshighereducation.com/impactrankings/industry-innovation-and-infrastructure', 'https://www.timeshighereducation.com/impactrankings/reducing-inequalities'
, 'https://www.timeshighereducation.com/impactrankings/sustainable-cities-and-communities', 'https://www.timeshighereducation.com/impactrankings/responsible-consumption-and-production', 'https://www.timeshighereducation.com/impactrankings/climate-action', 
 'https://www.timeshighereducation.com/impactrankings/life-below-water', 'https://www.timeshighereducation.com/impactrankings/life-land', 'https://www.timeshighereducation.com/impactrankings/peace-justice-and-strong-institutions'
, 'https://www.timeshighereducation.com/impactrankings/partnerships-goals']
sdglist[0]

'https://www.timeshighereducation.com/impactrankings'

### structure
- Rank : Name : No poverty : Overall
<table id='datatable-1'>
    <thead>
        <tr>
            <th class='rank sroting_asc'>Rank </th>
            <th class='name namesearch sorting'>Name </th>
            <th class='sdg-scores sorting'>No Poverty</th>
            <th class='scores overall-score sorting'>Overall</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td class='rank sorting_1 sorting_2'>1</td>
            <td class=' name namesearch'>
                <a class='ranking-institution-title' href='/world-university-rankings/universiti-sains-malaysia'>Universiti Sains Malaysia</a>
                <div class='location'> <span> <a href='/location/a4zw000000GNwUAAS'>Malaysia</a></span> </div> </td>
            <td class=' sdg-scores'>
                <div class='sdg-score'>94.3</div></td>
            <td class=' scores overall-score'>95.6</td>
        </tr>
    </tbody>
</table>

In [88]:
#start driver
cService = webdriver.ChromeService(executable_path=cp2)
driver = webdriver.Chrome(service = cService)
driver.get(sdglist[0])
wait = WebDriverWait(driver, 3)
print(driver.title)

University Impact Rankings for UN SDG 1: no poverty | Times Higher Education (THE)


In [89]:
# Maximize window and wait for page to load
#driver.get(sdglist[0])
driver.maximize_window()
time.sleep(3)  
driver.find_element(By.ID,'CybotCookiebotDialogBodyButtonDecline').click()

In [90]:
#check if table element exists
table = driver.find_element(By.ID,'datatable-1')
table

<selenium.webdriver.remote.webelement.WebElement (session="1f616297e6e6e8b1893ff7b2559efec4", element="f.8B7151914B16B26B234911E8C4424F04.d.601D2A6F6AF8538322EC5C67986526D1.e.1593")>

In [44]:
# Locate the table header row
header_elements = driver.find_elements(By.CSS_SELECTOR, '#datatable-1 thead th')
# Extract and print header text
headers = [header.text.strip() for header in header_elements]
print(headers)

['Rank', 'Name', 'No poverty', 'Overall']


In [55]:
# Step 1: Find all rows inside tbody
rows = driver.find_elements(By.CSS_SELECTOR, "table#datatable-1 tbody tr")
rows[0:5] #if 5 elements are shown

[<selenium.webdriver.remote.webelement.WebElement (session="5f5c4f5fc1927330a899ec337f20c3c1", element="f.55D63FD87B1AF39AAD83F624D366A63A.d.F6DCB01AF49B3BFCCF8AA6FC55E2573D.e.1778")>,
 <selenium.webdriver.remote.webelement.WebElement (session="5f5c4f5fc1927330a899ec337f20c3c1", element="f.55D63FD87B1AF39AAD83F624D366A63A.d.F6DCB01AF49B3BFCCF8AA6FC55E2573D.e.1779")>,
 <selenium.webdriver.remote.webelement.WebElement (session="5f5c4f5fc1927330a899ec337f20c3c1", element="f.55D63FD87B1AF39AAD83F624D366A63A.d.F6DCB01AF49B3BFCCF8AA6FC55E2573D.e.1780")>,
 <selenium.webdriver.remote.webelement.WebElement (session="5f5c4f5fc1927330a899ec337f20c3c1", element="f.55D63FD87B1AF39AAD83F624D366A63A.d.F6DCB01AF49B3BFCCF8AA6FC55E2573D.e.1781")>,
 <selenium.webdriver.remote.webelement.WebElement (session="5f5c4f5fc1927330a899ec337f20c3c1", element="f.55D63FD87B1AF39AAD83F624D366A63A.d.F6DCB01AF49B3BFCCF8AA6FC55E2573D.e.1782")>]

In [56]:
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
len(rows2)

25

In [61]:
part2 ="#!/length/-1/sort_by/rank/sort_order/asc"


'https://www.timeshighereducation.com/impactrankings/no-poverty#!/length/-1/sort_by/rank/sort_order/asc'

### SDG00 Overall

In [135]:
sdg00 = sdglist[0] + part2
sdg00
driver.get(sdg00)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
len(rows2) #2232

2232

In [137]:
#SDG Overall
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
len(rows2) #2232
data_sdg00 = []
i, sdg=0, 0
for row in rows2:
    i = i + 1
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_container = row.find_element(By.CSS_SELECTOR, "td.best-scores .sdg-score-multi")
    sdg_items = sdg_container.find_elements(By.CLASS_NAME, "sdg-score-multi__item")
    sdg_scores = {}
    for item in sdg_items:
        number = item.find_element(By.CLASS_NAME, "sdg-score-multi__number").text.strip()
        value = item.find_element(By.CLASS_NAME, "sdg-score-multi__value p").text.strip()
        sdg_scores[number] = value
        overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg00.append({'ser':i,  "rank": rank, "university": name, "country": country, 'sdgno':sdg, "sdg_scores": sdg_scores, "overall_score": overall })
    
print("No of rows extracted ", i, ' for overall summary')
#----------
flat_data = []
for entry in data_sdg00:
    base = {k: v for k, v in entry.items() if k != 'sdg_scores'}
    if 'sdg_scores' in entry:
        for i, (sdg_no, sdg_val) in enumerate(entry['sdg_scores'].items(), start=1):
            base[f'SDG{i}_no'] = sdg_no
            base[f'SDG{i}_value'] = sdg_val
    flat_data.append(base)

df_sdg00 = pd.DataFrame(flat_data)
print(df_sdg00.shape)

InvalidSessionIdException: Message: invalid session id: session deleted as the browser has closed the connection
from disconnected: Unable to receive message from renderer
  (Session info: chrome=137.0.7151.120)
Stacktrace:
0   chromedriver                        0x0000000104d5a654 cxxbridge1$str$ptr + 2723108
1   chromedriver                        0x0000000104d528c8 cxxbridge1$str$ptr + 2690968
2   chromedriver                        0x00000001048a6714 cxxbridge1$string$len + 90428
3   chromedriver                        0x0000000104891604 cxxbridge1$string$len + 4140
4   chromedriver                        0x000000010489137c cxxbridge1$string$len + 3492
5   chromedriver                        0x000000010488f134 chromedriver + 192820
6   chromedriver                        0x000000010488fcc4 chromedriver + 195780
7   chromedriver                        0x00000001048abf78 cxxbridge1$string$len + 113056
8   chromedriver                        0x00000001048ae710 cxxbridge1$string$len + 123192
9   chromedriver                        0x00000001048ae7b8 cxxbridge1$string$len + 123360
10  chromedriver                        0x00000001048ed390 cxxbridge1$string$len + 380344
11  chromedriver                        0x00000001048e2f94 cxxbridge1$string$len + 338364
12  chromedriver                        0x000000010492ede8 cxxbridge1$string$len + 649232
13  chromedriver                        0x00000001048e19c8 cxxbridge1$string$len + 332784
14  chromedriver                        0x0000000104d1e278 cxxbridge1$str$ptr + 2476360
15  chromedriver                        0x0000000104d2150c cxxbridge1$str$ptr + 2489308
16  chromedriver                        0x0000000104cffa64 cxxbridge1$str$ptr + 2351412
17  chromedriver                        0x0000000104d21d94 cxxbridge1$str$ptr + 2491492
18  chromedriver                        0x0000000104cf0d58 cxxbridge1$str$ptr + 2290728
19  chromedriver                        0x0000000104d41d60 cxxbridge1$str$ptr + 2622512
20  chromedriver                        0x0000000104d41eec cxxbridge1$str$ptr + 2622908
21  chromedriver                        0x0000000104d52514 cxxbridge1$str$ptr + 2690020
22  libsystem_pthread.dylib             0x000000018da56c0c _pthread_start + 136
23  libsystem_pthread.dylib             0x000000018da51b80 thread_start + 8


### SDG01 No Poverty

In [96]:
sdg01 = sdglist[1] + part2
sdg01
driver.get(sdg01)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
len(rows2) #1222

1222

In [63]:
sdg_score = rows2[0].find_element(By.CLASS_NAME, "sdg-score").text.strip()
sdg_score

'94.3'

In [64]:
data_sdg01 = []
i, sdg=0, 1
#print(rows[0:1])
for row in rows2:
    i = i + 1
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    sdg=1
    data_sdg01.append({'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i)

No of rows extracted  1222


In [65]:
df_sdg01 = pd.DataFrame(data_sdg01)
df_sdg01.shape

(1222, 6)

### SDG02 Zero Hunger

In [76]:
#SDG02
sdg02 = sdglist[2] + part2
print(sdg02)
driver.get(sdg02)
wait = WebDriverWait(driver, 3)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2))

https://www.timeshighereducation.com/impactrankings/zero-hunger#!/length/-1/sort_by/rank/sort_order/asc
921


In [78]:
data_sdg02 = []
i, sdg=0, 2
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg02.append({'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i)

No of rows extracted  921


In [80]:
df_sdg02 = pd.DataFrame(data_sdg02)
df_sdg02.shape

(921, 6)

In [ ]:
### SDG03 Good Health

In [98]:
#SDG03
sdg03 = sdglist[3] + part2
print(sdg03)
driver.get(sdg03)
wait = WebDriverWait(driver, 3)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2)) # 1719

https://www.timeshighereducation.com/impactrankings/good-health-and-well-being#!/length/-1/sort_by/rank/sort_order/asc
1719


In [99]:
data_sdg03 = []
i, sdg=0, 3
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg03.append({ 'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i)

No of rows extracted  1719


In [75]:
df_sdg03 = pd.DataFrame(data_sdg03)
df_sdg03.shape

(1889, 6)

In [79]:
df_sdg03.head()

rank                     university    country  sdgno sdg_score  \
0    1   Lingnan University Hong Kong  Hong Kong      3      93.3   
1    2   Hong Kong Baptist University  Hong Kong      3      92.2   
2   =3             Aalborg University    Denmark      3      91.1   
3   =3  Istanbul Technical University     Turkey      3      91.1   
4   =5    Al-Ahliyya Amman University     Jordan      3      89.8   

  overall_score  
0     83.7–89.4  
1          91.2  
2          96.9  
3          92.6  
4     83.7–89.4

### SDG04 Quality Education

In [81]:
#SDG04-------------
sdg04 = sdglist[4] + part2
print(sdg04)
driver.get(sdg04)
wait = WebDriverWait(driver, 3)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2))
#------
data_sdg04 = []
i, sdg=0, 4
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg04.append({ 'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i)
#-------
df_sdg04 = pd.DataFrame(data_sdg04)
df_sdg04.shape

https://www.timeshighereducation.com/impactrankings/quality-education#!/length/-1/sort_by/rank/sort_order/asc
1889
No of rows extracted  1889


(1889, 6)

### SDG05 Quality Education

In [82]:
#SDG05 -------------
sdg05 = sdglist[5] + part2
print(sdg05)
driver.get(sdg05)
wait = WebDriverWait(driver, 3)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2))
#------
data_sdg05 = []
i, sdg=0, 5
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg05.append({'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i)
#-------
df_sdg05 = pd.DataFrame(data_sdg05)
df_sdg05.shape

https://www.timeshighereducation.com/impactrankings/gender-equality#!/length/-1/sort_by/rank/sort_order/asc
1508
No of rows extracted  1508


(1508, 6)

In [85]:
#df_sdg05['sdgno'] = 5
df_sdg05.head()

rank               university    country  sdgno sdg_score overall_score
0    1      Walailak University   Thailand      5      85.0          89.7
1    2    Bahçeşehir University     Turkey      5      82.8          91.6
2    3  University of Bucharest    Romania      5      82.6          89.7
3    4    Universitas Airlangga  Indonesia      5      82.1          96.9
4   =5      Flinders University  Australia      5      81.6          95.0

### SDG06 Clean Water

In [91]:
#SDG06-------------
sdg06 = sdglist[6] + part2
print(sdg06)
driver.get(sdg06)
wait = WebDriverWait(driver, 3)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2))
#------
data_sdg06 = []
i, sdg=0, 6
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg06.append({'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i)
#-------
df_sdg06 = pd.DataFrame(data_sdg06)
df_sdg06.shape

https://www.timeshighereducation.com/impactrankings/clean-water-and-sanitation#!/length/-1/sort_by/rank/sort_order/asc
1008
No of rows extracted  1008


(1008, 6)

### SDG07 Clean Energy

In [101]:
#SDG07-------------
sdg07 = sdglist[7] + part2
print(sdg07)
driver.get(sdg07)
wait = WebDriverWait(driver, 3)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2))
#------
data_sdg07 = []
i, sdg=0, 7
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg07.append({'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i)
#-------
df_sdg07 = pd.DataFrame(data_sdg07)
df_sdg07.shape #1146

https://www.timeshighereducation.com/impactrankings/affordable-and-clean-energy#!/length/-1/sort_by/rank/sort_order/asc
1146
No of rows extracted  1146


(1146, 6)

### SDG08 Decent Work

In [102]:
#SDG08-------------
sdg08 = sdglist[8] + part2
print(sdg08)
driver.get(sdg08)
wait = WebDriverWait(driver, 3)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2))
#------
data_sdg08 = []
i, sdg=0, 8
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg08.append({'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i, 'for sdg ', sdg)
#-------
df_sdg08 = pd.DataFrame(data_sdg08)
df_sdg08.shape

https://www.timeshighereducation.com/impactrankings/decent-work-and-economic-growth#!/length/-1/sort_by/rank/sort_order/asc
1286
No of rows extracted  1286


(1286, 6)

### SDG09 Industry

In [104]:
#SDG09-------------
sdg09 = sdglist[9] + part2
print(sdg09)
driver.get(sdg09)
wait = WebDriverWait(driver, 4)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2))
#------
data_sdg09 = []
i, sdg=0, 9
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg09.append({'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i, 'for sdg ', sdg)
#-------
df_sdg09 = pd.DataFrame(data_sdg09)
df_sdg09.shape

https://www.timeshighereducation.com/impactrankings/industry-innovation-and-infrastructure#!/length/-1/sort_by/rank/sort_order/asc
1096
No of rows extracted  1096 for sdg  9


(1096, 6)

### SDG10 Equality

In [106]:
#SDG10-------------
sdg10 = sdglist[10] + part2
print(sdg10)
driver.get(sdg10)
wait = WebDriverWait(driver, 4)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2))
#------
data_sdg10 = []
i, sdg=0, 10
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg10.append({'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i, 'for sdg ', sdg)
#-------
df_sdg10 = pd.DataFrame(data_sdg10)
df_sdg10.shape

https://www.timeshighereducation.com/impactrankings/reducing-inequalities#!/length/-1/sort_by/rank/sort_order/asc
1213
No of rows extracted  1213 for sdg  10


(1213, 6)

### SDG11 Sustainable Cities

In [108]:
#SDG11-------------
sdg11 = sdglist[11] + part2
print(sdg11)
driver.get(sdg11)
wait = WebDriverWait(driver, 4)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2))
#------
data_sdg11 = []
i, sdg=0, 11
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg11.append({'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i, 'for sdg ', sdg)
#-------
df_sdg11 = pd.DataFrame(data_sdg11)
df_sdg11.shape

https://www.timeshighereducation.com/impactrankings/sustainable-cities-and-communities#!/length/-1/sort_by/rank/sort_order/asc
1101
No of rows extracted  1101 for sdg  11


(1101, 6)

### SDG12 Consumption

In [131]:
#SDG12-------------
sdg12 = sdglist[12] + part2
print(sdg12)
driver.get(sdg12)
wait = WebDriverWait(driver, 4)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2))
#------
data_sdg12 = []
i, sdg=0, 12
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg12.append({'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i, 'for sdg ', sdg)
#-------
df_sdg12 = pd.DataFrame(data_sdg12)
df_sdg12.shape

https://www.timeshighereducation.com/impactrankings/responsible-consumption-and-production#!/length/-1/sort_by/rank/sort_order/asc
940
No of rows extracted  940 for sdg  12


(940, 6)

### SDG13 Climate

In [130]:
#SDG13-------------
sdg13 = sdglist[13] + part2
print(sdg13)
driver.get(sdg13)
wait = WebDriverWait(driver, 4)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2))
#------
data_sdg13 = []
i, sdg=0, 13
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg13.append({'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i, 'for sdg ', sdg)
#-------
df_sdg13 = pd.DataFrame(data_sdg13)
df_sdg13.shape

https://www.timeshighereducation.com/impactrankings/climate-action#!/length/-1/sort_by/rank/sort_order/asc
1054
No of rows extracted  1054 for sdg  13


(1054, 6)

### SDG14 Water

In [129]:
#SDG14-------------
sdg14 = sdglist[14] + part2
print(sdg14)
driver.get(sdg14)
wait = WebDriverWait(driver, 4)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2))
#------
data_sdg14 = []
i, sdg=0, 14
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg14.append({'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i, 'for sdg ', sdg)
#-------
df_sdg14 = pd.DataFrame(data_sdg14)
df_sdg14.shape

https://www.timeshighereducation.com/impactrankings/life-below-water#!/length/-1/sort_by/rank/sort_order/asc
682
No of rows extracted  682 for sdg  14


(682, 6)

### SDG15 Life on Land

### SDG16 Peace

In [128]:
#SDG15-------------
sdg15 = sdglist[15] + part2
print(sdg15)
driver.get(sdg15)
wait = WebDriverWait(driver, 4)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2))
#------
data_sdg15 = []
i, sdg=0, 15
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg15.append({'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i, 'for sdg ', sdg)
#-------
df_sdg15 = pd.DataFrame(data_sdg15)
df_sdg15.shape

https://www.timeshighereducation.com/impactrankings/life-land#!/length/-1/sort_by/rank/sort_order/asc
820
No of rows extracted  820 for sdg  15


(820, 6)

### SDG17 Partnerships

In [118]:
#SDG16-------------
sdg16 = sdglist[16] + part2
print(sdg16)
driver.get(sdg16)
wait = WebDriverWait(driver, 4)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2))
#------
data_sdg16 = []
i, sdg=0, 16
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg16.append({'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i, 'for sdg ', sdg)
#-------
df_sdg16 = pd.DataFrame(data_sdg16)
df_sdg16.shape

https://www.timeshighereducation.com/impactrankings/peace-justice-and-strong-institutions#!/length/-1/sort_by/rank/sort_order/asc
1214
No of rows extracted  1214 for sdg  16


(1214, 6)

In [116]:
#SDG17-------------
sdg17 = sdglist[17] + part2
print(sdg17)
driver.get(sdg17)
wait = WebDriverWait(driver, 4)
rows2 = driver.find_elements(By.XPATH, "//tr[contains(@class, 'js-row') and not(contains(@class, 'institution-disabled'))]")
print(len(rows2))
#------
data_sdg17 = []
i, sdg=0, 17
for row in rows2:
    i = i + 1 
    rank = row.find_element(By.CSS_SELECTOR, "td.rank.sorting_1.sorting_2").text.strip()
    name = row.find_element(By.CSS_SELECTOR, "td.name a.ranking-institution-title").text.strip()
    country = row.find_element(By.CSS_SELECTOR, "td.name .location span a").text.strip()
    sdg_score = row.find_element(By.CLASS_NAME, "sdg-score").text.strip()
    overall = row.find_element(By.CSS_SELECTOR, "td.scores.overall-score").text.strip()
    data_sdg17.append({'ser':i, "rank": rank, "university": name, "country": country, 'sdgno': sdg, "sdg_score": sdg_score, "overall_score": overall })
print("No of rows extracted ", i, 'for sdg ', sdg)
#-------
df_sdg17 = pd.DataFrame(data_sdg17)
df_sdg17.shape

https://www.timeshighereducation.com/impactrankings/partnerships-goals#!/length/-1/sort_by/rank/sort_order/asc
2299
No of rows extracted  2299 for sdg  17


(2299, 6)

### All DFs
- Shape of multiple DF if they exist

In [138]:
listDF = ['df_sdg00','df_sdg01','df_sdg02','df_sdg03', 'df_sdg04','df_sdg05','df_sdg06','df_sdg07','df_sdg08','df_sdg09','df_sdg10','df_sdg11','df_sdg12','df_sdg13','df_sdg14','df_sdg15','df_sdg16','df_sdg17']
for name in listDF:
    try:
        df = eval(name)
        print(f"{name}: {df.shape}")
    except NameError:
        continue

df_sdg00: (2232, 13)
df_sdg01: (1222, 6)
df_sdg02: (921, 6)
df_sdg03: (1889, 6)
df_sdg04: (1889, 6)
df_sdg05: (1508, 6)
df_sdg06: (1008, 6)
df_sdg07: (1146, 6)
df_sdg08: (1286, 6)
df_sdg09: (1096, 6)
df_sdg10: (1213, 6)
df_sdg11: (1101, 6)
df_sdg12: (940, 6)
df_sdg13: (1054, 6)
df_sdg14: (682, 6)
df_sdg15: (820, 6)
df_sdg16: (1214, 6)
df_sdg17: (2299, 6)


In [139]:
file_path = '/Users/du/dup/auData/ranking/the/impact25.xlsx'

with pd.ExcelWriter(file_path, engine='xlsxwriter') as writer:
    for name in listDF:
        if name in globals():
            globals()[name].to_excel(writer, sheet_name=name, index=False)

print(f"Saved to {file_path}")

Saved to /Users/du/dup/auData/ranking/the/impact25.xlsx


In [140]:
# Load the Excel file
excel_file = pd.ExcelFile(file_path)
# List all sheet names
print(excel_file.sheet_names)

['df_sdg00', 'df_sdg01', 'df_sdg02', 'df_sdg03', 'df_sdg04', 'df_sdg05', 'df_sdg06', 'df_sdg07', 'df_sdg08', 'df_sdg09', 'df_sdg10', 'df_sdg11', 'df_sdg12', 'df_sdg13', 'df_sdg14', 'df_sdg15', 'df_sdg16', 'df_sdg17']
